<a href="https://colab.research.google.com/github/DavidSenseman/BIO1173_Fall2025/blob/main/AI_Scientist_06.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

---------------------------
**COPYRIGHT NOTICE:** This Jupyterlab Notebook is a Derivative work of [Jeff Heaton](https://github.com/jeffheaton) licensed under the Apache License, Version 2.0 (the "License"); You may not use this file except in compliance with the License. You may obtain a copy of the License at

> [http://www.apache.org/licenses/LICENSE-2.0](http://www.apache.org/licenses/LICENSE-2.0)

Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.

------------------------

# **BIO 1173: Intro Computational Biology**

##### **Module 6: Advanced Topics**

* Instructor: [David Senseman](mailto:David.Senseman@utsa.edu), [Department of Biology, Health and the Environment](https://sciences.utsa.edu/bhe/), [UTSA](https://www.utsa.edu/)

### Module 6 Material

* Part 6.1: Reenforcement Learning
* **Part 6.2: AI-Scientist**
* Part 6.3: Generative AI
* Part 6.4: Text to Images with Stable Diffusion


## Google CoLab Instructions

You MUST run the following code cell to get credit for this class lesson. By running this code cell, you will map your GDrive to /content/drive and print out your Google GMAIL address. Your Instructor will use your GMAIL address to verify the author of this class lesson.

In [1]:
# You must run this cell first
try:
    from google.colab import drive
    drive.mount('/content/drive', force_remount=True)
    # from google.colab import auth
    # auth.authenticate_user()
    COLAB = True
    print("Note: Using Google CoLab")
    # import requests
    # gcloud_token = !gcloud auth print-access-token
    # gcloud_tokeninfo = requests.get('https://www.googleapis.com/oauth2/v3/tokeninfo?access_token=' + gcloud_token[0]).json()
    # print(gcloud_tokeninfo['email'])
except:
    print("**WARNING**: Your GMAIL address was **not** printed in the output below.")
    print("**WARNING**: You will NOT receive credit for this lesson.")
    COLAB = False

Mounted at /content/drive
Note: Using Google CoLab


Make sure your GMAIL address is included as the last line in the output above.

# **AI-Scientist Project by SakanaAI**

The **AI-Scientist project** by SakanaAI is an ambitious initiative aimed at automating the entire scientific research process. It leverages advanced **Large Language Models (LLMs)** and other AI technologies to independently conduct research, from generating ideas to writing full scientific papers.



Get Key

In [ ]:
# Retrieve the OpenAI API key and store it in a variable
from google.colab import userdata

# Retrieve the OpenAI API key and store it in a variable
OPENAI_API_KEY = userdata.get('OPENAI_KEY')

# Ensure that the API key is correctly set
if not OPENAI_API_KEY:
    raise ValueError("OpenAI API key is not set. Please check if you have stored the API key in userdata.")

In [ ]:
!pip install backoff
!pip install anthropic
!pip install openai


In [ ]:
# !pip install git+https://github.com/Aider-AI/aider.git

## **Install GitHub AI-Scientist**

In [2]:

!apt-get install git -y
!git clone https://github.com/SakanaAI/AI-Scientist.git
%cd AI-Scientist
!cp -r /content/AI-Scientist/* /content/
!ls /content/


Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
git is already the newest version (1:2.34.1-1ubuntu1.12).
0 upgraded, 0 newly installed, 0 to remove and 29 not upgraded.
Cloning into 'AI-Scientist'...
remote: Enumerating objects: 2638, done.
remote: Counting objects: 100% (10/10), done.
remote: Compressing objects: 100% (9/9), done.
remote: Total 2638 (delta 3), reused 1 (delta 1), pack-reused 2628 (from 2)
Receiving objects: 100% (2638/2638), 116.87 MiB | 41.62 MiB/s, done.
Resolving deltas: 100% (464/464), done.
Updating files: 100% (2319/2319), done.
/content/AI-Scientist
ai_scientist  docs	      experimental	   README.md		review_iclr_bench
AI-Scientist  drive	      launch_scientist.py  requirements.txt	sample_data
data	      example_papers  LICENSE		   review_ai_scientist	templates


~~~text
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
git is already the newest version (1:2.34.1-1ubuntu1.12).
0 upgraded, 0 newly installed, 0 to remove and 29 not upgraded.
Cloning into 'AI-Scientist'...
remote: Enumerating objects: 2638, done.
remote: Counting objects: 100% (10/10), done.
remote: Compressing objects: 100% (9/9), done.
remote: Total 2638 (delta 3), reused 1 (delta 1), pack-reused 2628 (from 2)
Receiving objects: 100% (2638/2638), 116.87 MiB | 34.45 MiB/s, done.
Resolving deltas: 100% (464/464), done.
Updating files: 100% (2319/2319), done.
/content/AI-Scientist/AI-Scientist
ai_scientist  docs	      experimental	   README.md		review_iclr_bench
AI-Scientist  drive	      launch_scientist.py  requirements.txt	sample_data
data	      example_papers  LICENSE		   review_ai_scientist	templates

~~~

In [6]:
!ls -Fla /content/AI-Scientist/templates/mobilenetV3/

total 60
drwxr-xr-x  3 root root  4096 Mar 25 20:15 ./
drwxr-xr-x 12 root root  4096 Mar 25 20:15 ../
-rw-r--r--  1 root root 21966 Mar 25 20:15 experiment.py
-rw-r--r--  1 root root  1135 Mar 25 20:15 ideas.json
drwxr-xr-x  2 root root  4096 Mar 25 20:15 latex/
-rw-r--r--  1 root root  8752 Mar 25 20:15 plot.py
-rw-r--r--  1 root root   425 Mar 25 20:15 prompt.json
-rw-r--r--  1 root root  1119 Mar 25 20:15 seed_ideas.json


In [ ]:
!mkdir /content/data
!mkdir /content/data/mobilenetV3
!mkdir /content/data/results/
!mkdir /content/data/results/mobilenetV3
!ls -Fla /content/data/results

!python

# **Create Jason Idea File**

In [18]:
# Create Idea file

import json
import os

# Example data (replace this with your actual "ideas" data)
ideas = {
      "Name": "CNN analysis of non-coding genetic variants",
      "Title": "Identifying Functional Non-Coding Variants",
      "Experiment": "Use CNNs to predict the regulatory roles of non-coding genetic variants identified in GWAS.",
      "Interestingness": 9,
      "Feasibility": 7,
      "Novelty": 8
}

# Define the target directory
output_dir = "/content/AI-Scientist/templates/mobilenetV3/"
os.makedirs(output_dir, exist_ok=True)  # Ensure the directory exists

# Define the full path for the output file
output_file = os.path.join(output_dir, "seed_ideas.json")

# Save the data as a JSON file
with open(output_file, "w") as json_file:
    json.dump(ideas, json_file, indent=4)

print(f"Your ideas have been saved to {output_file}!")


Your ideas have been saved to /content/AI-Scientist/templates/mobilenetV3/seed_ideas.json!


In [20]:
import json

# Path to your JSON file (adjust the path as needed)
file_path = '/content/AI-Scientist/templates/mobilenetV3/seed_ideas.json'

# Open and load the JSON file
with open(file_path, 'r') as json_file:
    data = json.load(json_file)

# Print the contents of the JSON file
print(data)



{'Name': 'CNN analysis of non-coding genetic variants', 'Title': 'Identifying Functional Non-Coding Variants', 'Experiment': 'Use CNNs to predict the regulatory roles of non-coding genetic variants identified in GWAS.', 'Interestingness': 9, 'Feasibility': 7, 'Novelty': 8}


# **Create Jason Prompt File**

In [21]:
# Create Idea file

import json
import os

# Example data (replace this with your actual "ideas" data)
prompt = {
    "system": "You are an ambitious AI PhD student who is looking to publish a paper that will contribute significantly to the field.",
    "task_description": "You are given the following file to work with, which studies Convolutional neural networks by training CNN models on multiple GWAS datasets. Please come up with interesting experiments to investigate the best architectures for these convolutional networks."
}

# Define the target directory
output_dir = "/content/AI-Scientist/templates/mobilenetV3/"
os.makedirs(output_dir, exist_ok=True)  # Ensure the directory exists

# Define the full path for the output file
output_file = os.path.join(output_dir, "prompt.json")

# Save the data as a JSON file
with open(output_file, "w") as json_file:
    json.dump(ideas, json_file, indent=4)

print(f"Your ideas have been saved to {output_file}!")


Your ideas have been saved to /content/AI-Scientist/templates/mobilenetV3/prompt.json!


In [14]:
!ls -Fla /content/AI-Scientist/templates/mobilenetV3/

total 60
drwxr-xr-x  3 root root  4096 Mar 25 20:15 ./
drwxr-xr-x 12 root root  4096 Mar 25 20:15 ../
-rw-r--r--  1 root root 21966 Mar 25 20:15 experiment.py
-rw-r--r--  1 root root  1135 Mar 25 20:15 ideas.json
drwxr-xr-x  2 root root  4096 Mar 25 20:15 latex/
-rw-r--r--  1 root root  8752 Mar 25 20:15 plot.py
-rw-r--r--  1 root root   299 Mar 25 20:23 prompt.json
-rw-r--r--  1 root root   299 Mar 25 20:19 seed_ideas.json


In [22]:
import json

# Path to your JSON file (adjust the path as needed)
file_path = '/content/AI-Scientist/templates/mobilenetV3/prompt.json'

# Open and load the JSON file
with open(file_path, 'r') as json_file:
    data = json.load(json_file)

# Print the contents of the JSON file
print(data)



{'Name': 'CNN analysis of non-coding genetic variants', 'Title': 'Identifying Functional Non-Coding Variants', 'Experiment': 'Use CNNs to predict the regulatory roles of non-coding genetic variants identified in GWAS.', 'Interestingness': 9, 'Feasibility': 7, 'Novelty': 8}


## **Run Experiment**

In [ ]:
!python /content/AI-Scientist/templates/mobilenetV3/experiment.py

# **Generate Ideas - Step 1**

In [ ]:
!ls -Fla /content/AI-Scientist/AI-Scientist/ai_scientist

total 12
drwxr-xr-x 3 root root 4096 Mar 24 21:41 ./
drwxr-xr-x 3 root root 4096 Mar 24 21:41 ../
drwxr-xr-x 3 root root 4096 Mar 24 21:41 templates/
